hazelcast/src/main/java/com/hazelcast/query/impl/getters/AbstractJsonGetter.java

```diff
@@ -0,0 +1,196 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.query.impl.getters;
+
+import com.fasterxml.jackson.core.JsonParser;
+import com.fasterxml.jackson.core.JsonToken;
+import com.fasterxml.jackson.core.JsonTokenId;
+import com.hazelcast.core.HazelcastException;
+import com.hazelcast.internal.json.JsonValue;
+import com.hazelcast.nio.serialization.HazelcastSerializationException;
+
+import java.io.IOException;
+
+public abstract class AbstractJsonGetter extends Getter {
+
+    private static final String DELIMITER = "\\.|\\[";
+
+    public AbstractJsonGetter(Getter parent) {
+        super(parent);
+    }
+
+    public static Object convertFromJsonValue(JsonValue value) {
+        if (value == null) {
+            return null;
+        } else if (value.isNumber()) {
+            try {
+                return value.asLong();
+            } catch (NumberFormatException e) {
+                return value.asDouble();
+            }
+        } else if (value.isBoolean()) {
+            return value.asBoolean();
+        } else if (value.isNull()) {
+            return null;
+        } else if (value.isString()) {
+            return value.asString();
+        }
+        throw new HazelcastSerializationException("Unknown Json type: " + value);
+    }
+
+    protected String[] getPath(String attributePath) {
+        return attributePath.split(DELIMITER);
+    }
+
+    @Override
+    Object getValue(Object obj) throws Exception {
+        throw new HazelcastException("Path agnostic value extraction is not supported");
+
+    }
+
+    @Override
+    Object getValue(Object obj, String attributePath) throws Exception {
+        String[] paths = getPath(attributePath);
+        JsonParser parser = createParser(obj);
+
+        parser.nextToken();
+        for (int i = 0; i < paths.length; i++) {
+            String path = paths[i];
+            String arrayIndexText = getIndexTextOrNull(path);
+            if (arrayIndexText == null) {
+                // non array case
+
+
+                if (!findAttribute(parser, path)) {
+                    return null;
+                }
+            } else {
+                // array case
+                if ("any".equals(arrayIndexText)) {
+                    String lastPath = i + 1 < paths.length ? paths[i + 1] : null;
+                    return getMultiValue(parser, lastPath);
+                } else {
+                    JsonToken token = parser.getCurrentToken();
```

Comment:
Minor: maybe you can use `.currentToken()` since the javadoc says:
```
    /**
     * Alias for {@link #currentToken()}, will be deprecated in Jackson 2.9
     */
```

Reply:
Minor: there are still some other uses of `parser.get*` which are going to be deprecated.